airflow/executors/kubernetes_executor.py

```diff
@@ -62,6 +64,50 @@
 KubernetesWatchType = Tuple[str, str, Optional[str], Dict[str, str], str]
 
 
+def multi_threads_queue_process(
+    queue_size: int,
+    queue_type: str,
+    process_method: Callable,
+    max_threads: int,
+    log: Logger,
+    batch_size: Optional[int] = None,
+) -> None:
+    """
+    Helper method to enable multi-threads for processing queues used with kubernetes executor
+    :param queue_size: the size of the queue getting processed
+    :param queue_type: the type of the queue
+    :param process_method: the real method processing the queue
+    :param max_threads: the max num of threads to be used
+    :param log: log
+    :param batch_size: the max num of items we want to process in this round.
+                       If it's not set, the current queue size will be used.
+    """
+    if queue_size == 0:
+        log.info(f'There is no item to process in the {queue_type} queue.')
+        return
+
+    start_time = time.time()
+    log.info(f'Start processing {queue_type} queue with at most {max_threads} threads.')
+
+    batch_size = min(batch_size or queue_size, queue_size)
+    max_threads = min(max_threads, queue_size)
+
+    threads = []
+    quotient, remainder = divmod(batch_size, max_threads)
+    for i in range(max_threads):
+        sub_batch_size = quotient + 1 if i < remainder else quotient
+        t = Thread(target=process_method, args=[sub_batch_size])
+        threads.append(t)
+        t.start()
+    for t in threads:
+        t.join()
```

Comment:
This seems like it's more complicated than necessary.

I just played around with threadpoolexecutor and it seems it can be a bit simpler.  E.g. here's a simple example:
```
import time
from concurrent.futures import ThreadPoolExecutor


def print_and_sleep(x):
    print(x)
    time.sleep(0.25)

max_threads = 4
with ThreadPoolExecutor(max_threads) as pool:
    [pool.submit(print_and_sleep, x) for x in range(123)]
```

Reply:
ok i guess we can do something like this to see what failed and handle appropriately:
```
import time
from concurrent.futures import ThreadPoolExecutor


def print_and_sleep(x):
    print(x)
    if x % 50 == 0:
        raise ValueError("hi")
    time.sleep(0.25)


max_threads = 4
with ThreadPoolExecutor(max_threads) as pool:
    futures = [(x, pool.submit(print_and_sleep, x)) for x in range(123)]

failures = []
for arg, f in futures:
    print(arg)
    if f.exception():
        failures.append((arg, f))

for arg, f in failures:
    print(f"Func print_and_sleep failed for arg {arg} with exception {f.exception()}")

```

Reply:
Hi @dstandish , thanks for the reply. The idea is the same. IIRC, I was using ThreadPoolExecutor at the very beginning and then met some issues in production with more workloads. Unfortunately, did not record the issue. But with the current implementation, there was no issue and it has more flexibility for us to do updates. So would like to still go with that.

Reply:
if you want to do this manually, at least i would separate the batch construction logic from the thread creation logic e.g. you could do this

```
def get_batches(
    queue_size=100,
    batch_size=10,
    max_threads=3,
):
    batch_size = min(batch_size or queue_size, queue_size)
    max_threads = min(max_threads, queue_size)
    quotient, remainder = divmod(batch_size, max_threads)
    for i in range(max_threads):
        sub_batch_size = quotient + 1 if i < remainder else quotient
        yield sub_batch_size

threads = []
for num in get_batches():
    t = Thread(target=process_method, args=[num])
    threads.append(t)
    t.start()

for t in threads:
    t.join()

```

or something

it's not obvious what this batch definition logic is supposed to do

Reply:
(and splitting means you can test that part of it pretty easily)

Reply:
better yet...

why have fixed batches?  why not just launch desired number of threads, and each thread can consume queue until the queue is consumed. no need to manage batches right?

Reply:
oh i see... worker_pods_creation_batch_size was already there... it limits num to create each loop

Reply:
but you see the issue i'm thinking of year? since you are defining fixed sizes for each thread, one thread may finish more quickly than the others, so it could do more work if the others are taking longer per-item.

so it would be more efficient if each thread just drew from the same queue.

BUT the complication here is that you need to limit the total number of tasks processed because of that setting.  i think one way to handle that is you could create a new queue object for each batch with all items taken from task_queue that you need to process in in the batch.  but maybe this is not really gonna matter too much.  and we needn't let perfect be enemy of good.

other thing i'll mention is....  k8s has support for async requests... not asyncio, but threads... it manages its own thread pool... you use by adding `async_req=True` to request.  thread count is configured in client.  but refactoring to this woud be much more involved so not suggesting it, but just want to mention.

anyway, i'll keep watching this one closely.  i have been a bit busy over last month with holidays and illness and my personal projects but i'm excited that y'all are looking to get more involved and am committed to helping you see this one through and encouraging the any others coming in the future. thanks.